src/channel_internal.c

```diff
@@ -0,0 +1,395 @@
+{
+    int result;
+    /*Codes_SRS_CHANNEL_INTERNAL_43_103: [ channel_internal_pull shall create a THANDLE(ASYNC_OP) by calling async_op_create with cancel_op as cancel. ]*/
+    /*Codes_SRS_CHANNEL_INTERNAL_43_119: [ channel_internal_push shall create a THANDLE(ASYNC_OP) by calling async_op_create with cancel_op as cancel. ]*/
+    THANDLE(ASYNC_OP) async_op = async_op_create(cancel_op, sizeof(CHANNEL_OP), alignof(CHANNEL_OP), dispose_channel_op);
+    if (async_op == NULL)
+    {
+        LogError("Failure in async_op_create(cancel_op, sizeof(CHANNEL_OP), alignof(CHANNEL_OP), dispose_channel_op)");
+        result = MU_FAILURE;
+    }
+    else
+    {
+        CHANNEL_OP* channel_op = async_op->context;
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_104: [ channel_internal_pull shall store the pull_callback and pull_context in the THANDLE(ASYNC_OP). ]*/
+        channel_op->pull_callback = pull_callback;
+        channel_op->pull_context = pull_context;
+
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_120: [ channel_internal_push shall store the push_callback, push_context and data in the THANDLE(ASYNC_OP). ]*/
+        channel_op->push_callback = push_callback;
+        channel_op->push_context = push_context;
+
+        THANDLE_INITIALIZE(ASYNC_OP)(&channel_op->async_op, async_op);
+        THANDLE_INITIALIZE(CHANNEL_INTERNAL)(&channel_op->channel_internal, channel_internal);
+        THANDLE_INITIALIZE(RC_PTR)(&channel_op->data, data);
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_111: [ channel_internal_pull shall set the result of the created operation to CHANNEL_CALLBACK_RESULT_OK. ]*/
+        /*Codes_SRS_CHANNEL_INTERNAL_43_127: [ channel_internal_push shall set the result of the created operation to CHANNEL_CALLBACK_RESULT_OK. ]*/
+        channel_op->result = CHANNEL_CALLBACK_RESULT_OK;
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_105: [ channel_internal_pull shall insert the created THANDLE(ASYNC_OP) in the list of pending operations by calling DList_InsertTailList. ]*/
+        /*Codes_SRS_CHANNEL_INTERNAL_43_121: [ channel_internal_push shall insert the created THANDLE(ASYNC_OP) in the list of pending operations by calling DList_InsertTailList. ]*/
+        DList_InsertTailList(&channel_internal->op_list, &channel_op->anchor);
+
+        /*Codes_SRS_CHANNEL_INTERNAL_43_107: [ channel_internal_pull shall set *out_op_pull to the created THANDLE(ASYNC_OP). ]*/
+        /*Codes_SRS_CHANNEL_INTERNAL_43_123: [ channel_internal_push shall set *out_op_push to the created THANDLE(ASYNC_OP). ]*/
+        THANDLE_INITIALIZE_MOVE(ASYNC_OP)(out_op, &async_op);
+        result = 0;
+    }
+    return result;
+}
+
+static int dequeue_operation(CHANNEL_INTERNAL* channel_internal, THANDLE(ASYNC_OP)* out_op, PULL_CALLBACK pull_callback, void* pull_context, PUSH_CALLBACK push_callback, void* push_context, THANDLE(RC_PTR) data)
```

Comment:
>dequeue_operation [](http://example.com/codeflow?start=11&length=17)

Is this function ever called with both PUSH/PULL callbacks set to non-NULL values?

Reply:
No

Reply:
`dequeue_operation` does the symmetric "other half" of work as `enqueue_operation`